src/main/java/com/cleanroommc/groovyscript/compat/mods/artisanworktables/Tables.java

```diff
@@ -0,0 +1,458 @@
+                    this.registryName,
+                    data.width,
+                    data.height,
+                    this.tools.size(),
+                    this.input.size(),
+                    this.fluidInput.isEmpty() ? null : this.fluidInput.get(0)
+            );
+            // weird cornercase that the mod doesn't handle
+            if (!shaped && data.ingredients.size() > 9 && tier != null) {
+                if (data.ingredients.size() <= 25) tier = EnumTier.WORKSHOP;
+                else tier = null;
+            }
+            GroovyLog.Msg message =
+                    GroovyLog.msg("Error adding Artisan Worktables recipe").error()
+                    .add(tier == null, "Unable to calculate the table tier");
+            if (tier != null) {
+                int tierNeeded = tier.getId();
+                message.add(tierNeeded > maximumTier, "The recipe requires at least tier {}, but maximumTier was set to {}", tierNeeded, maximumTier);
+                minimumTier = Math.max(minimumTier, tierNeeded);
+            }
+            if (message.postIfNotEmpty()) return null;
+
+            // The mod expects three Nonnull ExtraOutputChancePairs... yuck
+            ExtraOutputChancePair[] extraOutputs = new ExtraOutputChancePair[3];
+            Arrays.fill(extraOutputs, new ExtraOutputChancePair(ArtisanItemStack.EMPTY, 0));
+            for (int i = 0; i < this.extraOutputs.size(); i++) extraOutputs[i] = this.extraOutputs.get(i);
+
+            IArtisanRecipe recipe = new ArtisanRecipe(
+                    this.makeArtisanName(),
+                    requirements,
+                    output,
+                    tools.toArray(new ToolEntry[0]),
+                    data.ingredients.stream().map(r -> ArtisanIngredient.from(r.toMcIngredient())).collect(Collectors.toList()),
+                    input.stream().map(r -> ArtisanIngredient.from(r.toMcIngredient())).collect(Collectors.toList()),
+                    consumeSecondaryIngredients,
+                    fluidInput.isEmpty() ? null : fluidInput.get(0),
+                    experience,
+                    level,
+                    consumeExperience,
+                    extraOutputs,
+                    shaped ? IRecipeMatrixMatcher.SHAPED : IRecipeMatrixMatcher.SHAPELESS,
+                    mirrored,
+                    data.width,
+                    data.height,
+                    minimumTier,
+                    maximumTier,
+                    hidden);
+            ModSupport.ARTISAN_WORKTABLES.get().tables.add(recipe);
+            return recipe;
+        }
+
+        @Override
+        public String getErrorMsg() {
+            return "Error adding Artisan Worktable recipe";
+        }
+
+        @Override
+        public void validate(GroovyLog.Msg msg) {
+            if (name == null || name.isEmpty()) {
+                name = RecipeName.generate("groovyscript_");
+            }
+
+            validateItems(msg, 0, 9, 0, 0);

```

Comment:
i think there needs to be at least one input, right?

Reply:
the inputs is for *secondary* inputs, not for the grid inputs. Each composited class validates grid inputs itself.